File: conversation.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
)

// Read the actual message ledger, including tool receipts. Never manufacture
// missing results or silently omit an unfinished tool batch from the context.
func conversationMessages(state string, parent *Run) ([]Message, error) {
	raw, err := readRunFile(state, parent.ID, "session.jsonl")
	if err != nil {
		return nil, errors.New("上一轮完整消息记录不可用，无法继续；请开始新任务")
	}
	messages := []Message{}
	pending := map[string]bool{}
	users := 0
	invalid := errors.New("上一轮消息记录不完整或工具回执未配对，无法继续；请开始新任务")
	for _, line := range bytes.Split(bytes.TrimSpace(raw), []byte("\n")) {
		var record struct {
			Type    string  `json:"type"`
			Message Message `json:"message"`
		}
		if json.Unmarshal(line, &record) != nil || record.Type != "message" {
			return nil, invalid
		}
		m := record.Message
		if len(messages) == 0 && m.Role != "system" {
			return nil, invalid
		}
		if len(pending) > 0 && m.Role != "tool" {
			return nil, invalid
		}
		if m.Role != "assistant" && len(m.ToolCalls) > 0 || m.Role != "tool" && m.ToolCallID != "" {
			return nil, invalid
		}
		switch m.Role {
		case "system":
			if len(messages) != 0 {
				return nil, invalid
			}
		case "user":
			users++
		case "assistant":
			for _, call := range m.ToolCalls {
				if call.ID == "" || pending[call.ID] || call.Type != "function" || call.Function.Name == "" {
					return nil, invalid
				}
				pending[call.ID] = true
			}
		case "tool":
			if !pending[m.ToolCallID] {
				return nil, invalid
			}
			delete(pending, m.ToolCallID)
		default:
			return nil, invalid
		}
		messages = append(messages, m)
	}
	if len(pending) > 0 || users != max(1, parent.ConversationTurn) {
		return nil, invalid
	}
	return messages, nil
}

File: src/conversation.ts
import type { Message } from './agent.ts';
import { parseMessage } from './model.ts';
import { object, readRunFile } from './storage.ts';
import type { Run } from './storage.ts';
export function conversationMessages(state: string, parent: Run): Message[] {
    const invalid = new Error('上一轮消息记录不完整或工具回执未配对，无法继续；请开始新任务');
    let raw: Buffer;
    try {
        raw = readRunFile(state, parent.id, 'session.jsonl');
    }
    catch {
        throw new Error('上一轮完整消息记录不可用，无法继续；请开始新任务');
    }
    const messages: Message[] = [], pending = new Set<string>();
    let users = 0;
    try {
        for (const line of raw.toString().trim().split('\n')) {
            const record = object(JSON.parse(line));
            if (record.type !== 'message')
                throw invalid;
            const m = parseMessage(record.message);
            if (!messages.length && m.role !== 'system' || pending.size && m.role !== 'tool')
                throw invalid;
            if (m.role !== 'assistant' && m.tool_calls?.length || m.role !== 'tool' && m.tool_call_id)
                throw invalid;
            switch (m.role) {
                case 'system':
                    if (messages.length)
                        throw invalid;
                    break;
                case 'user':
                    users++;
                    break;
                case 'assistant':
                    for (const call of m.tool_calls ?? []) {
                        if (!call.id || pending.has(call.id) || call.type !== 'function' || !call.function.name)
                            throw invalid;
                        pending.add(call.id);
                    }
                    break;
                case 'tool':
                    if (!m.tool_call_id || !pending.delete(m.tool_call_id))
                        throw invalid;
                    break;
                default: throw invalid;
            }
            messages.push(m);
        }
        if (pending.size || users !== parent.conversation_turn)
            throw invalid;
    }
    catch {
        throw invalid;
    }
    return messages;
}
